anomalib/data/avenue.py

```diff
@@ -0,0 +1,289 @@
+    """Create CUHK Avenue dataset by parsing the file structure.
+
+    The files are expected to follow the structure:
+        path/to/dataset/[training_videos|testing_videos]/video_filename.avi
+        path/to/ground_truth/mask_filename.mat
+
+    Args:
+        root (Path): Path to dataset
+        gt_dir (Path): Path to the ground truth
+        split (Optional[Union[Split, str]], optional): Dataset split (ie., either train or test). Defaults to None.
+
+    Example:
+        The following example shows how to get testing samples from Avenue dataset:
+
+        >>> root = Path('./avenue')
+        >>> gt_dir = Path('./avenue/masks')
+        >>> samples = make_avenue_dataset(path, gt_dir, split='test')
+        >>> samples.head()
+           root     folder         image_path                      mask_path                   split
+        0  ./avenue testing_videos ./avenue/training_videos/01.avi ./avenue/masks/01_label.mat test
+        1  ./avenue testing_videos ./avenue/training_videos/02.avi ./avenue/masks/01_label.mat test
+        ...
+
+    Returns:
+        DataFrame: an output dataframe containing samples for the requested split (ie., train or test)
+    """
+
+    samples_list = [(str(root),) + filename.parts[-2:] for filename in Path(root).glob("**/*.avi")]
+    samples = DataFrame(samples_list, columns=["root", "folder", "image_path"])
+
+    samples.loc[samples.folder == "testing_videos", "mask_path"] = (
+        samples.image_path.str.split(".").str[0].str.lstrip("0") + "_label.mat"
+    )
+    samples.loc[samples.folder == "testing_videos", "mask_path"] = (
+        str(gt_dir) + "/testing_label_mask/" + samples.mask_path
+    )
+    samples.loc[samples.folder == "training_videos", "mask_path"] = ""
+
+    samples["image_path"] = samples.root + "/" + samples.folder + "/" + samples.image_path
+
+    samples.loc[samples.folder == "training_videos", "split"] = "train"
+    samples.loc[samples.folder == "testing_videos", "split"] = "test"
+
+    if split:
+        samples = samples[samples.split == split]
+        samples = samples.reset_index(drop=True)
+
+    return samples
+
+
+class AvenueClipsIndexer(ClipsIndexer):
+    """Clips class for UCSDped dataset."""
+
+    def get_mask(self, idx) -> Optional[Tensor]:
+        """Retrieve the masks from the file system."""
+
+        video_idx, frames_idx = self.get_clip_location(idx)
+        matfile = self.mask_paths[video_idx]
+        if matfile == "":  # no gt masks available for this clip
+            return None
+        frames = self.clips[video_idx][frames_idx]
+
+        # read masks from .png files if available, othwerise from mat files.
+        mask_folder = Path(matfile).with_suffix("")
+        if mask_folder.exists():
+            mask_frames = sorted(glob.glob(str(mask_folder) + "/*"))
+            mask_paths = [mask_frames[idx] for idx in frames.int()]
+            masks = torch.stack([Tensor(cv2.imread(mask_path, flags=0)) for mask_path in mask_paths])
```

Comment:
This was to keep the masks consistent with the video frames, which are also retrieved as Tensor, as dictated by the torchvision's VideoClips class. But since it leads to 2 unnecessary conversions, it might be better to retrieve the masks as numpy array actually.

Reply:
done